clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/ttl/VeniceRmdTTLFilter.java

```diff
@@ -27,68 +32,121 @@
+    this.valueDeserializerCache = new VeniceConcurrentHashMap<>();
+    this.rmdSerializerCache = new VeniceConcurrentHashMap<>();
+    this.valueSerializerCache = new VeniceConcurrentHashMap<>();
   }
 
   @Override
-  public boolean apply(final INPUT_VALUE value) {
+  public boolean checkAndMaybeFilterValue(final INPUT_VALUE value) {
     if (skipRmdRecord(value)) {
       return false;
     }
-    Instant curTime = Instant.now();
-    switch (ttlPolicy) {
-      case RT_WRITE_ONLY:
-        Instant timestamp = Instant.ofEpochMilli(getTimeStampFromRmdRecord(value));
-        return ChronoUnit.MILLIS.between(timestamp, curTime) > ttlInMs;
-      default:
-        throw new UnsupportedOperationException(ttlPolicy + " policy is not supported.");
+    if (Objects.requireNonNull(ttlPolicy) == TTLResolutionPolicy.RT_WRITE_ONLY) {
+      return validateByTTLandMaybeUpdateValue(value, filterTimestamp);
     }
+    throw new UnsupportedOperationException(ttlPolicy + " policy is not supported.");
   }
 
   @Override
   public void close() {
     schemaSource.close();
   }
 
-  public long getTimeStampFromRmdRecord(final INPUT_VALUE value) {
+  boolean validateByTTLandMaybeUpdateValue(final INPUT_VALUE value, long filterTimestamp) {
     ByteBuffer rmdPayload = getRmdPayload(value);
     if (rmdPayload == null || !rmdPayload.hasRemaining()) {
       throw new IllegalStateException(
           "The record doesn't contain required RMD field. Please check if your store has A/A enabled");
     }
-    int id = getRmdId(value), valueSchemaId = getSchemaId(value);
+    int valueSchemaId = getSchemaId(value);
+    int id = getRmdProtocolId(value);
     RmdVersionId rmdVersionId = new RmdVersionId(valueSchemaId, id);
-    GenericRecord record =
-        this.rmdDeserializerCache.computeIfAbsent(rmdVersionId, this::generateDeserializer).deserialize(rmdPayload);
-    return RmdUtils.extractTimestampFromRmd(record)
-        .stream()
-        .mapToLong(v -> v)
-        .max()
-        .orElseThrow(NoSuchElementException::new);
+    GenericRecord rmdRecord =
+        rmdDeserializerCache.computeIfAbsent(rmdVersionId, this::generateRmdDeserializer).deserialize(rmdPayload);
+    Object rmdTimestampObject = rmdRecord.get(TIMESTAMP_FIELD_POS);
+    RmdTimestampType rmdTimestampType = RmdUtils.getRmdTimestampType(rmdTimestampObject);
+    // For value-level RMD timestamp, just compare the value with the filter TS.
+    if (rmdTimestampType.equals(RmdTimestampType.VALUE_LEVEL_TIMESTAMP)) {
+      return (long) rmdTimestampObject < filterTimestamp;
+    }
+    RecordDeserializer<GenericRecord> valueDeserializer =
+        valueDeserializerCache.computeIfAbsent(valueSchemaId, this::generateValueDeserializer);
+    GenericRecord valueRecord = valueDeserializer.deserialize(getValuePayload(value));
+    UpdateResultStatus updateResultStatus =
+        mergeRecordHelper.deleteRecord(valueRecord, (GenericRecord) rmdTimestampObject, filterTimestamp - 1, 0);
```

Comment:
may be define `filterTimestamp` as filterTimestamp -1 and update the line 93 logic as such ?

Reply:
OK

Reply:
Yeah. Fixed it.